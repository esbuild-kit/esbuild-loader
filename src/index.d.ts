import { EsbuildPluginOptions } from './types';

export class EsbuildPlugin {
	constructor(options?: EsbuildPluginOptions);

	apply(compiler: any): void;
}

export * from './types';
