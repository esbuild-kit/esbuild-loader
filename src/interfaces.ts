import { transform, TransformOptions } from 'esbuild';

type Filter = string | RegExp;

type Implementation = {
	transform: typeof transform;
};

type TransformImports = {
	[key: string]: {
		transform: (importName:string, matches:RegExpMatchArray) => string | string;
		preventFullImport?: boolean;
		skipDefaultConversion?: boolean;
	};
}

type Except<ObjectType, Properties> = {
	[Key in keyof ObjectType as (Key extends Properties ? never : Key)]: ObjectType[Key];
};

type LoaderOptions = Except<TransformOptions, 'sourcemap' | 'sourcefile'> & {
	/** Pass a custom esbuild implementation */
	implementation?: Implementation;
	transformImports?: TransformImports;
};

type MinifyPluginOptions = Except<TransformOptions, 'sourcefile'> & {
	include?: Filter | Filter[];
	exclude?: Filter | Filter[];
	css?: boolean;
	/** Pass a custom esbuild implementation */
	implementation?: Implementation;
};

export {
	LoaderOptions,
	TransformImports,
	MinifyPluginOptions,
};
