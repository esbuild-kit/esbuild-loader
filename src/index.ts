import esbuildLoader from './loader';
import ESBuildPlugin from './plugin';
import ESBuildMinifyPlugin from './minify-plugin';

/* eslint-disable unicorn/prefer-export-from */
export default esbuildLoader;
export { ESBuildPlugin, ESBuildMinifyPlugin };
/* eslint-enable unicorn/prefer-export-from */
