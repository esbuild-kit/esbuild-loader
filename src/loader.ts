import path from 'path';
import {
	transform as defaultEsbuildTransform,
	type TransformOptions,
} from 'esbuild';
import { getOptions } from 'loader-utils';
import webpack from 'webpack';
import {
	getTsconfig,
	parseTsconfig,
	createFilesMatcher,
	type TsConfigResult,
} from 'get-tsconfig';
import type { LoaderOptions } from './types.js';

const tsconfigCache = new Map<string, TsConfigResult>();

async function ESBuildLoader(
	this: webpack.loader.LoaderContext<LoaderOptions>,
	source: string,
): Promise<void> {
	const done = this.async()!;
	const options: LoaderOptions = typeof this.getOptions === 'function' ? this.getOptions() : getOptions(this);
	const {
		implementation,
		tsconfig: tsconfigPath,
		...esbuildTransformOptions
	} = options;

	if (implementation && typeof implementation.transform !== 'function') {
		done(
			new TypeError(
				`esbuild-loader: options.implementation.transform must be an ESBuild transform function. Received ${typeof implementation.transform}`,
			),
		);
		return;
	}

	const transform = implementation?.transform ?? defaultEsbuildTransform;

	const transformOptions = {
		...esbuildTransformOptions,
		target: options.target ?? 'es2015',
		loader: options.loader ?? 'default',
		sourcemap: this.sourceMap,
		sourcefile: this.resourcePath,
	};

	if (!('tsconfigRaw' in transformOptions)) {
		const { resourcePath } = this;

		/**
		 * If a tsconfig.json path is specified, force apply it
		 * Same way a provided tsconfigRaw is applied regardless
		 * of whether it actually matches
		 *
		 * However in this case, we also warn if it doesn't match
		 */
		if (tsconfigPath) {
			const tsconfigFullPath = path.resolve(tsconfigPath);
			const cacheKey = `esbuild-loader:${tsconfigFullPath}`;
			let tsconfig = tsconfigCache.get(cacheKey);
			if (!tsconfig) {
				tsconfig = {
					config: parseTsconfig(tsconfigFullPath, tsconfigCache),
					path: tsconfigFullPath,
				};
				tsconfigCache.set(cacheKey, tsconfig);
			}

			const filesMatcher = createFilesMatcher(tsconfig);
			const matches = filesMatcher(resourcePath);

			if (!matches) {
				this.emitWarning(
					new Error(`[esbuild-loader] The specified tsconfig at "${tsconfigFullPath}" was applied to the file "${resourcePath}" but does not match its "include" patterns`),
				);
			}

			transformOptions.tsconfigRaw = tsconfig.config as TransformOptions['tsconfigRaw'];
		} else {
			/* Detect tsconfig file */

			// Webpack shouldn't be loading the same path multiple times so doesn't need to be cached
			const tsconfig = getTsconfig(resourcePath, 'tsconfig.json', tsconfigCache);
			if (tsconfig) {
				const fileMatcher = createFilesMatcher(tsconfig);
				transformOptions.tsconfigRaw = fileMatcher(resourcePath) as TransformOptions['tsconfigRaw'];
			}
		}
	}

	try {
		const { code, map } = await transform(source, transformOptions);
		done(null, code, map && JSON.parse(map));
	} catch (error: unknown) {
		done(error as Error);
	}
}

export default ESBuildLoader;
