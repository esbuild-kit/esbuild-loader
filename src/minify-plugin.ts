import { transform as defaultEsbuildTransform } from 'esbuild';
import { RawSource, SourceMapSource } from 'webpack-sources';
import webpack from 'webpack';
import type {
	SyncHook, SyncBailHook, AsyncSeriesHook, HookMap,
} from 'tapable';
import type { Source } from 'webpack-sources';
import { matchObject } from 'webpack/lib/ModuleFilenameHelpers.js';
import { MinifyPluginOptions } from './interfaces';

type StatsPrinter = {
	hooks: {
		print: HookMap<SyncBailHook<any, string>>;
	};
};

type Compilation = webpack.compilation.Compilation;

type Wp5Compilation = Compilation & {
	hooks: Compilation['hooks'] & {
		processAssets: AsyncSeriesHook<Record<string, Source>>;
		statsPrinter: SyncHook<StatsPrinter>;
	};
	constructor: {
		PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE: 400;
	};
};

const isWebpack5 = (compilation: Compilation): compilation is Wp5Compilation => ('processAssets' in compilation.hooks);

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { version } = require('../package.json');

const isJsFile = /\.[cm]?js(\?.*)?$/i;
const isCssFile = /\.css(\?.*)?$/i;
const pluginName = 'esbuild-minify';

const granularMinifyConfigs = ['minifyIdentifiers', 'minifySyntax', 'minifyWhitespace'] as const;
class ESBuildMinifyPlugin {
	private readonly options: MinifyPluginOptions;

	private readonly transform: typeof defaultEsbuildTransform;

	constructor(options: MinifyPluginOptions = {}) {
		const { implementation, ...remainingOptions } = options;
		if (implementation && typeof implementation.transform !== 'function') {
			throw new TypeError(
				`ESBuildMinifyPlugin: implementation.transform must be an ESBuild transform function. Received ${typeof implementation.transform}`,
			);
		}

		this.transform = implementation?.transform ?? defaultEsbuildTransform;

		this.options = remainingOptions;

		const hasGranularMinificationConfig = granularMinifyConfigs.some(
			minifyConfig => minifyConfig in options,
		);

		if (!hasGranularMinificationConfig) {
			this.options.minify = true;
		}
	}

	apply(compiler: webpack.Compiler): void {
		const meta = JSON.stringify({
			name: 'esbuild-loader',
			version,
			options: this.options,
		});

		compiler.hooks.compilation.tap(pluginName, (compilation) => {
			compilation.hooks.chunkHash.tap(pluginName, (_, hash) => hash.update(meta));

			if (isWebpack5(compilation)) {
				compilation.hooks.processAssets.tapPromise(
					{
						name: pluginName,
						stage: compilation.constructor.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
						// @ts-expect-error TODO: modify type
						additionalAssets: true,
					},
					async () => await this.transformAssets(compilation),
				);

				compilation.hooks.statsPrinter.tap(pluginName, (statsPrinter) => {
					statsPrinter.hooks.print
						.for('asset.info.minimized')
						.tap(
							pluginName,
							(minimized, { green, formatFlag }) => (
								minimized
									? green(formatFlag('minimized'))
									: undefined
							),
						);
				});
			} else {
				compilation.hooks.optimizeChunkAssets.tapPromise(
					pluginName,
					async () => await this.transformAssets(compilation),
				);
			}
		});
	}

	private async transformAssets(
		compilation: Compilation,
	): Promise<void> {
		const { options: { devtool } } = compilation.compiler;

		const sourcemap = (
			// TODO: drop support for esbuild sourcemap in future so it all goes through WP API
			// Might still be necessary when SourceMap plugin is used
			this.options.sourcemap === undefined
				? devtool && (devtool as string).includes('source-map')
				: this.options.sourcemap
		);

		const {
			css: minifyCss,
			include,
			exclude,
			...transformOptions
		} = this.options;

		const assets = compilation.getAssets().filter(asset => (

			// Filter out already minimized
			!asset.info.minimized

			// Filter out by file type
			&& (
				isJsFile.test(asset.name)
				|| (minifyCss && isCssFile.test(asset.name))
			)
			&& matchObject({ include, exclude }, asset.name)
		));

		await Promise.all(assets.map(async (asset) => {
			const assetIsCss = isCssFile.test(asset.name);
			const { source, map } = asset.source.sourceAndMap();
			const sourceAsString = source.toString();
			const result = await this.transform(sourceAsString, {
				...transformOptions,
				loader: (
					assetIsCss
						? 'css'
						: transformOptions.loader
				),
				sourcemap,
				sourcefile: asset.name,
			});

			compilation.updateAsset(
				asset.name,
				(
					sourcemap
						? new SourceMapSource(
							result.code,
							asset.name,
							result.map as any,
							sourceAsString,
							map,
							true,
						)
						: new RawSource(result.code)
				),
				{
					...asset.info,
					minimized: true,
				},
			);
		}));
	}
}

export default ESBuildMinifyPlugin;
