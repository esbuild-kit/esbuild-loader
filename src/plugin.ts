import { transform as defaultEsbuildTransform } from 'esbuild';
import {
	RawSource as WP4RawSource,
	SourceMapSource as WP4SourceMapSource,
} from 'webpack-sources';
import webpack4 from 'webpack';
import webpack5 from 'webpack5';
import { matchObject } from 'webpack/lib/ModuleFilenameHelpers.js';
import { version } from '../package.json';
import type { EsbuildPluginOptions } from './types';

type Compiler = webpack4.Compiler | webpack5.Compiler;
type Compilation = webpack4.compilation.Compilation | webpack5.Compilation;
type Asset = webpack4.compilation.Asset | Readonly<webpack5.Asset>;

const isJsFile = /\.[cm]?js(?:\?.*)?$/i;
const isCssFile = /\.css(?:\?.*)?$/i;
const pluginName = 'esbuild-minify';

const granularMinifyConfigs = ['minifyIdentifiers', 'minifySyntax', 'minifyWhitespace'] as const;
class EsbuildPlugin {
	private readonly options: EsbuildPluginOptions;

	private readonly transform: typeof defaultEsbuildTransform;

	constructor(options: EsbuildPluginOptions = {}) {
		const {
			implementation,
			...remainingOptions
		} = options;

		if (implementation && typeof implementation.transform !== 'function') {
			throw new TypeError(
				`[${EsbuildPlugin.name}] implementation.transform must be an esbuild transform function. Received ${typeof implementation.transform}`,
			);
		}

		this.transform = implementation?.transform ?? defaultEsbuildTransform;

		const hasGranularMinificationConfig = granularMinifyConfigs.some(
			minifyConfig => minifyConfig in options,
		);

		if (!hasGranularMinificationConfig) {
			remainingOptions.minify = true;
		}

		this.options = remainingOptions;
	}

	apply(compiler: Compiler): void {
		const { options } = this;
		const meta = JSON.stringify({
			name: 'esbuild-loader',
			version,
			options,
		});

		if (!('format' in options)) {
			const { target } = compiler.options;
			const isWebTarget = (
				Array.isArray(target)
					? target.includes('web')
					: target === 'web'
			);
			const wontGenerateHelpers = !options.target || (
				Array.isArray(options.target)
					? (
						options.target.length === 1
						&& options.target[0] === 'esnext'
					)
					: options.target === 'esnext'
			);

			if (isWebTarget && !wontGenerateHelpers) {
				options.format = 'iife';
			}
		}

		compiler.hooks.compilation.tap(pluginName, (compilation) => {
			compilation.hooks.chunkHash.tap(pluginName, (_, hash) => hash.update(meta));

			/**
			 * Check if sourcemaps are enabled
			 * Webpack 4: https://github.com/webpack/webpack/blob/v4.46.0/lib/SourceMapDevToolModuleOptionsPlugin.js#L20
			 * Webpack 5: https://github.com/webpack/webpack/blob/v5.75.0/lib/SourceMapDevToolModuleOptionsPlugin.js#LL27
			 */
			let useSourceMap = false;
			compilation.hooks.finishModules.tap(
				pluginName,
				(modules) => {
					const firstModule = (
						Array.isArray(modules)
							? modules[0]
							: (modules as Set<webpack5.Module>).values().next().value as webpack5.Module
					);
					useSourceMap = firstModule.useSourceMap;
				},
			);

			if ('processAssets' in compilation.hooks) {
				compilation.hooks.processAssets.tapPromise(
					{
						name: pluginName,
						// @ts-expect-error undefined on Function type
						stage: compilation.constructor.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
						additionalAssets: true,
					},
					async () => await this.transformAssets(compilation, useSourceMap),
				);

				compilation.hooks.statsPrinter.tap(pluginName, (statsPrinter) => {
					statsPrinter.hooks.print
						.for('asset.info.minimized')
						.tap(
							pluginName,
							(
								minimized,
								{ green, formatFlag },
							// @ts-expect-error type incorrectly doesn't accept undefined
							) => (
								minimized
									// @ts-expect-error type incorrectly doesn't accept undefined
									? green(formatFlag('minimized'))
									: undefined
							),
						);
				});
			} else {
				compilation.hooks.optimizeChunkAssets.tapPromise(
					pluginName,
					async () => await this.transformAssets(compilation, useSourceMap),
				);
			}
		});
	}

	private async transformAssets(
		compilation: Compilation,
		useSourceMap: boolean,
	): Promise<void> {
		const { compiler } = compilation;
		const sources = 'webpack' in compiler && compiler.webpack.sources;
		const SourceMapSource = (sources ? sources.SourceMapSource : WP4SourceMapSource);
		const RawSource = (sources ? sources.RawSource : WP4RawSource);

		const {
			css: minifyCss,
			include,
			exclude,
			...transformOptions
		} = this.options;

		const assets = (compilation.getAssets() as Asset[]).filter(asset => (

			// Filter out already minimized
			!asset.info.minimized

			// Filter out by file type
			&& (
				isJsFile.test(asset.name)
				|| (minifyCss && isCssFile.test(asset.name))
			)
			&& matchObject({ include, exclude }, asset.name)
		));

		await Promise.all(assets.map(async (asset) => {
			const assetIsCss = isCssFile.test(asset.name);
			let source: string | Buffer | ArrayBuffer;
			let map = null;

			if (asset.source.sourceAndMap) {
				const sourceAndMap = asset.source.sourceAndMap();
				source = sourceAndMap.source;
				map = sourceAndMap.map;
			} else {
				source = asset.source.source();
				if (asset.source.map) {
					map = asset.source.map();
				}
			}

			const sourceAsString = source.toString();
			const result = await this.transform(sourceAsString, {
				...transformOptions,
				loader: (
					assetIsCss
						? 'css'
						: transformOptions.loader
				),
				sourcemap: useSourceMap,
				sourcefile: asset.name,
			});

			if (result.legalComments) {
				compilation.emitAsset(
					`${asset.name}.LEGAL.txt`,
					new RawSource(result.legalComments) as any,
				);
			}

			compilation.updateAsset(
				asset.name,
				(
					result.map
						? new SourceMapSource(
							result.code,
							asset.name,
							result.map as any,
							sourceAsString,
							map as any,
							true,
						)
						: new RawSource(result.code)
				) as any,
				{
					...asset.info,
					minimized: true,
				},
			);
		}));
	}
}

export default EsbuildPlugin;
