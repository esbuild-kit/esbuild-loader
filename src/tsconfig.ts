import { getTsconfig } from 'get-tsconfig';
import type { TransformOptions } from 'esbuild';

const foundTsconfig = getTsconfig();

export const tsconfig = foundTsconfig?.config as (TransformOptions['tsconfigRaw'] | undefined);
