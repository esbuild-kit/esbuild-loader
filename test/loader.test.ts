import webpack4 from 'webpack';
import webpack5 from 'webpack5';
import { build } from 'webpack-test-utils';
import { MinifyPluginOptions } from '../src/interfaces';
import * as fixtures from './fixtures';
import { configureEsbuildLoader } from './utils';

type WebpackSourceMapDevToolPlugin =
  | webpack4.SourceMapDevToolPlugin
  | webpack5.SourceMapDevToolPlugin;

describe.each([
	['Webpack 4', webpack4],
	['Webpack 5', webpack5],
])('%s', (_name, webpack) => {
	describe('Error handling', () => {
		test('tsx handled as ts', async () => {
			const built = await build(fixtures.tsx, (config) => {
				configureEsbuildLoader(config);

				config.module.rules.push({
					test: /\.tsx$/,
					loader: 'esbuild-loader',
					options: {
						loader: 'ts',
					},
				});
			}, webpack);

			expect(built.stats.hasErrors()).toBe(true);
			const [error] = built.stats.compilation.errors;
			expect(error.message).toMatch('Unexpected ">"');
		});

		test('invalid tsx', async () => {
			const built = await build(fixtures.invalidTsx, (config) => {
				configureEsbuildLoader(config);

				config.module.rules.push({
					test: /\.tsx?$/,
					loader: 'esbuild-loader',
					options: {
						loader: 'tsx',
					},
				});
			}, webpack);

			expect(built.stats.hasErrors()).toBe(true);
			const [error] = built.stats.compilation.errors;
			expect(error.message).toMatch('Unexpected "const"');
		});

		test('invalid implementation option', async () => {
			const runWithImplementation = async (implementation: MinifyPluginOptions['implementation']) => {
				const built = await build(fixtures.tsx, (config) => {
					configureEsbuildLoader(config);

					config.module.rules.push({
						test: /\.js?$/,
						loader: 'esbuild-loader',
						options: {
							implementation,
						},
					});
				}, webpack);

				expect(built.stats.hasErrors()).toBe(true);
				const [error] = built.stats.compilation.errors;
				throw error;
			};

			// @ts-expect-error testing invalid type
			await expect(runWithImplementation({})).rejects.toThrow(
				'esbuild-loader: options.implementation.transform must be an ESBuild transform function. Received undefined',
			);

			// @ts-expect-error testing invalid type
			await expect(runWithImplementation({ transform: 123 })).rejects.toThrow(
				'esbuild-loader: options.implementation.transform must be an ESBuild transform function. Received number',
			);
		});
	});

	describe('Loader', () => {
		test('js', async () => {
			const built = await build(fixtures.js, configureEsbuildLoader, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			expect(
				built.require('/dist'),
			).toMatchSnapshot();
		});

		test('ts', async () => {
			const built = await build(fixtures.ts, (config) => {
				configureEsbuildLoader(config);

				config.module.rules.push({
					test: /\.ts$/,
					loader: 'esbuild-loader',
					options: {
						loader: 'ts',
					},
				});
			}, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			expect(built.require('/dist')).toBe('foo');
		});

		test('tsx', async () => {
			const built = await build(fixtures.tsx, (config) => {
				configureEsbuildLoader(config);

				config.module.rules.push({
					test: /\.tsx$/,
					loader: 'esbuild-loader',
					options: {
						loader: 'tsx',
						jsxFactory: 'createElement',
						jsxFragment: 'Fragment',
					},
				});
			}, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			const dist = built.fs.readFileSync('/dist/index.js', 'utf-8');
			built.fs.writeFileSync(
				'/dist/index.js',
				`const createElement = (...args) => args, Fragment = "Fragment";${dist}`,
			);

			expect(
				built.require('/dist'),
			).toMatchSnapshot();
		});

		test('ts w/ tsconfig', async () => {
			const builtA = await build(fixtures.tsConfig, (config) => {
				configureEsbuildLoader(config);

				config.module.rules.push({
					test: /\.ts$/,
					loader: 'esbuild-loader',
					options: {
						loader: 'ts',
					},
				});
			}, webpack);

			expect(builtA.stats.hasWarnings()).toBe(false);
			expect(builtA.stats.hasErrors()).toBe(false);

			const distA = builtA.fs.readFileSync('/dist/index.js', 'utf-8');

			const builtB = await build(fixtures.tsConfig, (config) => {
				configureEsbuildLoader(config);

				config.module.rules.push({
					test: /\.ts$/,
					loader: 'esbuild-loader',
					options: {
						loader: 'ts',
						tsconfigRaw: {
							compilerOptions: {
								useDefineForClassFields: true,
							},
						},
					},
				});
			}, webpack);

			expect(builtB.stats.hasWarnings()).toBe(false);
			expect(builtB.stats.hasErrors()).toBe(false);

			const distB = builtB.fs.readFileSync('/dist/index.js', 'utf-8');
			expect(distB).not.toBe(distA);
		});

		test('tsx w/ tsconfig', async () => {
			const built = await build(fixtures.tsx, (config) => {
				configureEsbuildLoader(config);

				config.module.rules.push({
					test: /\.tsx$/,
					loader: 'esbuild-loader',
					options: {
						loader: 'tsx',
						tsconfigRaw: {
							compilerOptions: {
								jsxFactory: 'customFactory',
								jsxFragmentFactory: 'customFragment',
							},
						},
					},
				});
			}, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			const dist = built.fs.readFileSync('/dist/index.js', 'utf-8');
			built.fs.writeFileSync(
				'/dist/index.js',
				`const customFactory = (...args) => args, customFragment = "Fragment";${dist}`,
			);

			expect(
				built.require('/dist'),
			).toMatchSnapshot();
		});

		test('custom esbuild transform function', async () => {
			const built = await build(fixtures.ts, (config) => {
				configureEsbuildLoader(config);

				config.module.rules.push({
					test: /\.tsx?$/,
					loader: 'esbuild-loader',
					options: {
						loader: 'tsx',
						implementation: {
							transform: async () => ({
								code: 'export function foo() { return "MY_CUSTOM_ESBUILD_IMPLEMENTATION"; }',
								map: '',
								warnings: [],
							}),
						},
					},
				});
			}, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			const dist = built.fs.readFileSync('/dist/index.js', 'utf-8');

			expect(dist).toContain('MY_CUSTOM_ESBUILD_IMPLEMENTATION');
		});

		describe('ambigious ts/tsx', () => {
			test('ts via tsx', async () => {
				const built = await build(fixtures.ts, (config) => {
					configureEsbuildLoader(config);

					config.module.rules.push({
						test: /\.tsx?$/,
						loader: 'esbuild-loader',
						options: {
							loader: 'tsx',
						},
					});
				}, webpack);

				expect(built.stats.hasWarnings()).toBe(false);
				expect(built.stats.hasErrors()).toBe(false);

				expect(built.require('/dist')).toBe('foo');
			});

			test('ts via tsx 2', async () => {
				const built = await build(fixtures.ts2, (config) => {
					configureEsbuildLoader(config);

					config.module.rules.push({
						test: /\.tsx?$/,
						loader: 'esbuild-loader',
						options: {
							loader: 'tsx',
						},
					});
				}, webpack);

				expect(built.stats.hasWarnings()).toBe(false);
				expect(built.stats.hasErrors()).toBe(false);

				expect(
					built.require('/dist')('a', { a: 1 }),
				).toMatchSnapshot();
			});

			test('ambiguous ts', async () => {
				const built = await build(fixtures.tsAmbiguous, (config) => {
					configureEsbuildLoader(config);

					config.module.rules.push({
						test: /\.tsx?$/,
						loader: 'esbuild-loader',
						options: {
							loader: 'tsx',
						},
					});
				}, webpack);

				expect(built.stats.hasWarnings()).toBe(false);
				expect(built.stats.hasErrors()).toBe(false);

				const dist = built.fs.readFileSync('/dist/index.js', 'utf-8');
				expect(dist).toContain('(() => 1 < /a>/g)');
			});

			test('ambiguous tsx', async () => {
				const built = await build(fixtures.tsxAmbiguous, (config) => {
					configureEsbuildLoader(config);

					config.module.rules.push({
						test: /\.tsx?$/,
						loader: 'esbuild-loader',
						options: {
							loader: 'tsx',
						},
					});
				}, webpack);

				expect(built.stats.hasWarnings()).toBe(false);
				expect(built.stats.hasErrors()).toBe(false);

				const dist = built.fs.readFileSync('/dist/index.js', 'utf-8');
				expect(dist).toContain('React.createElement');
			});
		});
	});

	// Targets
	test('target', async () => {
		const built = await build(fixtures.js, (config) => {
			configureEsbuildLoader(config);

			config.module.rules[0].options = {
				target: 'es2015',
			};
		}, webpack);

		expect(built.stats.hasWarnings()).toBe(false);
		expect(built.stats.hasErrors()).toBe(false);

		expect(built.require('/dist')).toMatchSnapshot();
	});

	describe('Source-map', () => {
		test('source-map eval', async () => {
			const built = await build(fixtures.js, (config) => {
				configureEsbuildLoader(config);

				config.devtool = 'eval-source-map';
			}, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			const dist = built.fs.readFileSync('/dist/index.js', 'utf-8');
			expect(dist).toContain('eval');
		});

		test('source-map inline', async () => {
			const built = await build(fixtures.js, (config) => {
				configureEsbuildLoader(config);

				config.devtool = 'inline-source-map';
			}, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			const dist = built.fs.readFileSync('/dist/index.js', 'utf-8');
			expect(dist).toContain('sourceMappingURL');
		});

		test('source-map file', async () => {
			const built = await build(fixtures.js, (config) => {
				configureEsbuildLoader(config);

				config.devtool = 'source-map';
			}, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			const { assets } = built.stats.compilation;

			expect(assets).toHaveProperty(['index.js']);
			expect(assets).toHaveProperty(['index.js.map']);
		});

		test('source-map plugin', async () => {
			const built = await build(fixtures.js, (config) => {
				configureEsbuildLoader(config);

				delete config.devtool;
				(config.plugins as WebpackSourceMapDevToolPlugin[]).push(
					new webpack.SourceMapDevToolPlugin({}),
				);
			}, webpack);

			expect(built.stats.hasWarnings()).toBe(false);
			expect(built.stats.hasErrors()).toBe(false);

			const dist = built.fs.readFileSync('/dist/index.js', 'utf-8');
			expect(dist).toContain('sourceMappingURL');
		});
	});

	test('webpack magic comments', async () => {
		const built = await build(fixtures.webpackChunks, configureEsbuildLoader, webpack);

		expect(built.stats.hasWarnings()).toBe(false);
		expect(built.stats.hasErrors()).toBe(false);

		const { assets } = built.stats.compilation;

		expect(assets).toHaveProperty(['index.js']);
		expect(assets).toHaveProperty(['named-chunk-foo.js']);
		expect(assets).toHaveProperty(['named-chunk-bar.js']);
		expect(await built.require('/dist')()).toBe('foobar');
	});
});
