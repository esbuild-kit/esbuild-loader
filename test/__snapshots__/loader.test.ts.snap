// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webpack 4 Loader ambigious ts/tsx ts via tsx 2 1`] = `1`;

exports[`Webpack 4 Loader js 1`] = `
Object {
  "es2016": 10000,
  "es2017": "function",
  "es2018": Object {
    "a": 1,
  },
  "es2019": "try",
  "es2020": Array [
    1,
    2,
    Object {},
  ],
  "es2021": Array [
    2,
    3,
  ],
  "esnext": Array [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    1000000000,
  ],
}
`;

exports[`Webpack 4 Loader tsx 1`] = `
Array [
  Array [
    "Fragment",
    null,
    Array [
      "div",
      null,
      "hello world",
    ],
  ],
  Array [
    "div",
    Object {
      "className": "class-name",
    },
    "content",
  ],
]
`;

exports[`Webpack 4 Loader tsx w/ tsconfig 1`] = `
Array [
  Array [
    "Fragment",
    null,
    Array [
      "div",
      null,
      "hello world",
    ],
  ],
  Array [
    "div",
    Object {
      "className": "class-name",
    },
    "content",
  ],
]
`;

exports[`Webpack 4 target 1`] = `
Object {
  "es2016": 10000,
  "es2017": "function",
  "es2018": Object {
    "a": 1,
  },
  "es2019": "try",
  "es2020": Array [
    1,
    2,
    Object {},
  ],
  "es2021": Array [
    2,
    3,
  ],
  "esnext": Array [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    1000000000,
  ],
}
`;

exports[`Webpack 5 Loader ambigious ts/tsx ts via tsx 2 1`] = `1`;

exports[`Webpack 5 Loader js 1`] = `
Object {
  "es2016": 10000,
  "es2017": "function",
  "es2018": Object {
    "a": 1,
  },
  "es2019": "try",
  "es2020": Array [
    1,
    2,
    Object {},
  ],
  "es2021": Array [
    2,
    3,
  ],
  "esnext": Array [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    1000000000,
  ],
}
`;

exports[`Webpack 5 Loader tsx 1`] = `
Array [
  Array [
    "Fragment",
    null,
    Array [
      "div",
      null,
      "hello world",
    ],
  ],
  Array [
    "div",
    Object {
      "className": "class-name",
    },
    "content",
  ],
]
`;

exports[`Webpack 5 Loader tsx w/ tsconfig 1`] = `
Array [
  Array [
    "Fragment",
    null,
    Array [
      "div",
      null,
      "hello world",
    ],
  ],
  Array [
    "div",
    Object {
      "className": "class-name",
    },
    "content",
  ],
]
`;

exports[`Webpack 5 target 1`] = `
Object {
  "es2016": 10000,
  "es2017": "function",
  "es2018": Object {
    "a": 1,
  },
  "es2019": "try",
  "es2020": Array [
    1,
    2,
    Object {},
  ],
  "es2021": Array [
    2,
    3,
  ],
  "esnext": Array [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    1000000000,
  ],
}
`;
